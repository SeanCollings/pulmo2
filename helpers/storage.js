import { AsyncStorage } from 'react-native';
import { APP_ID } from '../constants/constants';

export const getAsyncData = async (id) => {
  try {
    const value = await AsyncStorage.getItem(`${APP_ID}${id}`);
    if (value !== null) {
      return JSON.parse(value);
    } else {
      return null;
    }
  } catch (error) {
    throw new Error(`Failure retrieving ${id} with error: ${error}`);
  }
};

export const getMultiAsyncData = async (id, keys) => {
  try {
    const formattedKeys = keys.map((key) => `${APP_ID}${id}${key}`);
    const values = await AsyncStorage.multiGet(formattedKeys);

    if (values && !!values.length) return values;
    else {
      return null;
    }
  } catch (error) {
    throw new Error(`Failure retrieving from all keys with error: ${error}`);
  }
};

export const storeAsyncData = async (id, value) => {
  try {
    await AsyncStorage.setItem(`${APP_ID}${id}`, JSON.stringify(value));
    return true;
  } catch (error) {
    throw new Error('Failure saving: ' + error);
  }
};

export const mergeAsyncData = async (id, value) => {
  try {
    await AsyncStorage.mergeItem(`${APP_ID}${id}`, JSON.stringify(value));
    return true;
  } catch (error) {
    throw new Error('Failure merging: ' + error);
  }
};

export const removeAsyncData = async (id) => {
  try {
    await AsyncStorage.removeItem(`${APP_ID}${id}`);
    return true;
  } catch (error) {
    throw new Error('Failure removing: ' + error);
  }
};
