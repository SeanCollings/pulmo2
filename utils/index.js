export const compose = (...fns) => (x) => fns.reduceRight((v, f) => f(v), x);

/**
 * Converts single digit to 2
 * @param {number} number
 */
export const formatNumber = (number) => `0${number}`.slice(-2);

/**
 * Converts milliseconds to seconds
 * @param {number} millis eg: 1234 milliseconds
 */
export const convertMilliToSeconds = (millis) => millis / 1000;

/**
 * Gets time in min:sec format
 * @param {number} time eg: 2 seconds
 */
export const getRemainingTime = (time) => {
  const mins = Math.floor(time / 60);
  const secs = Math.floor(time - mins * 60);
  return `${formatNumber(mins)}:${formatNumber(secs)}`;
};

/**
 * Converts milliseconds to min:sec format
 */
export const formatMilliToSeconds = compose(
  getRemainingTime,
  convertMilliToSeconds
);

/**
 * Returns an array start and end inclusive of given step values between
 * @param {number} start eg: 0
 * @param {number} end eg: 5
 * @param {number} step eg: 0.5
 */
export const getRangeArray = (start, end, step = 1) => {
  const len = Math.floor((end - start) / step) + 1;
  return Array(len)
    .fill()
    .map((_, idx) => start + idx * step);
};

/**
 * Converts a string in min:sec format to seconds
 * @param {string} time eg '02:12'
 */
export const convertToSeconds = (time) => {
  if (!time) return;
  const [min, sec] = time.split(':');
  const minToSec = +min * 60;
  return minToSec + +sec;
};

const appendZero = (input) => (input.length === 1 ? `0${input}` : input);

const getAmPmTime = (hours, mins) => {
  const ampm = hours < 12 || hours === 24 ? 'am' : 'pm';
  const hour = (hours % 12 || 12).toString();
  const ampPmTime = `${appendZero(hour)}:${mins} ${ampm}`;
  return ampPmTime;
};

/**
 * Converts to yyyy/mm/dd hh:mm format
 * @param {date} date eg 2020-12-06T14:15:00.015Z
 */
export const convertDate = (date) => {
  const newDate = new Date(date);
  const year = newDate.getFullYear().toString();
  const month = appendZero((newDate.getMonth() + 1).toString());
  const day = appendZero(newDate.getDate().toString());
  const hours = appendZero(newDate.getHours().toString());
  const mins = appendZero(newDate.getMinutes().toString());

  return `${year}/${month}/${day} ${hours}:${mins}`;
};

/**
 * Converts to DD MM dd @hh:mm
 * @param {date} date eg 2020-12-06T14:15:00.015Z
 */
export const convertDateDetail = (date) => {
  const newDate = new Date(date);

  const dayMonth = newDate.toDateString().substr(0, 8);
  const day = appendZero(newDate.getDate().toString());
  const hours = appendZero(newDate.getHours().toString());
  const mins = appendZero(newDate.getMinutes().toString());

  return `${dayMonth} ${day} @${hours}:${mins}`;
};

export const getDay = (date) => {
  const newDate = new Date(date);
  const dayMonth = newDate.toDateString().substr(0, 3);
  return dayMonth;
};

export const getTotalResultTime = (results) =>
  results.reduce((acc, result) => {
    const roundTime = result.reduce((accRound, round) => {
      if (!round) return accRound;
      return accRound + convertToSeconds(round);
    }, 0);
    return acc + roundTime;
  }, 0);
