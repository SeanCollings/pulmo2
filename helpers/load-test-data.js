import testData, {
  TEST_KEY,
  TEST_TITLE,
  TEST_MUTLIPLE_ID,
  TEST_DATA,
  TEST_PRE_ID,
} from '../data/test-data';
import { getAsyncData, storeAsyncData } from './storage';

export const MESSAGE_UNKNOWN = 'unknown';
export const MESSAGE_PRE_SAVED = 'pre-saved';
export const MESSAGE_DONE = 'done';

export default async (title) => {
  try {
    const selectedData = Object.values(testData).find(
      (data) => data[TEST_TITLE] === title
    );

    if (!selectedData) return MESSAGE_UNKNOWN;

    if (!selectedData[TEST_MUTLIPLE_ID]) {
      const savedData = await getAsyncData(selectedData[TEST_KEY]);

      if (!savedData) {
        console.log('Saving data for:', selectedData[TEST_KEY]);
        await storeAsyncData(selectedData[TEST_KEY], selectedData[TEST_DATA]);
        return MESSAGE_DONE;
      } else {
        return MESSAGE_PRE_SAVED;
      }
    } else {
      const updatedSaves = await selectedData[TEST_DATA].reduce(
        async (acc, data) => {
          const newAcc = await acc;
          const id = selectedData[TEST_MUTLIPLE_ID];
          const preId = selectedData[TEST_PRE_ID] || '';
          const savedData = await getAsyncData(`${preId}${data[id]}`);

          if (!savedData) {
            console.log('Saving data for:', `${preId}${data[id]}`);
            newAcc.push(data[id]);
            await storeAsyncData(`${preId}${data[id]}`, data);
          }
          return newAcc;
        },
        []
      );

      if (!updatedSaves.length) return MESSAGE_PRE_SAVED;

      return `${MESSAGE_DONE} [${updatedSaves.length}]`;
    }
  } catch (err) {
    throw new Error(err);
  }
};
