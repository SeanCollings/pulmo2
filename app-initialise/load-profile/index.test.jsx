import { cleanup } from 'react-native-testing-library';
import fixtures from '../__fixtures__/index.fixtures';
import { getMultiAsyncData } from '../../helpers/storage';
import loadProfileAsync from '.';
import { isDateToday } from '../../utils';

jest.mock('../../helpers/storage');
jest.mock('../../utils');

const lastActivityDate = '2020-07-26T15:15:43.411Z';
const profileResponse = [
  ['@Plumo2:selected_level', '2'],
  ['@Plumo2:streak', `{"current":0,"lastActivity":"${lastActivityDate}"}`],
  ['@Plumo2:disclaimer', null],
];
const profileResponseSameDay = [
  ['@Plumo2:selected_level', '2'],
  ['@Plumo2:streak', `{"current":2,"lastActivity":"${lastActivityDate}"}`],
  ['@Plumo2:disclaimer', null],
];
const profileResponseMissing = [
  ['@Plumo2:selected_level', '2'],
  ['@Plumo2:streak', `{"current":0,"lastActivity":"${lastActivityDate}"}`],
  ['@Plumo2:disclaimer', null],
  ['', ''],
  ['missing_value', ''],
  ['', 'missing_key'],
];
const profileNoStreakResponse = [
  ['@Plumo2:selected_level', '2'],
  ['@Plumo2:disclaimer', null],
];

describe('loadProfileAsync - unit test', () => {
  afterEach(cleanup);

  test('should return user profile', async () => {
    getMultiAsyncData.mockReturnValueOnce(profileResponse);

    const result = await loadProfileAsync();
    expect(result).toEqual(fixtures.profile);
  });

  test('should remove profiles without a key or value', async () => {
    getMultiAsyncData.mockReturnValueOnce(profileResponseMissing);

    const result = await loadProfileAsync();
    expect(result).toEqual(fixtures.profile);
  });

  test('should not set streak to 0 if last activity date is same day', async () => {
    getMultiAsyncData.mockReturnValueOnce(profileResponseSameDay);
    isDateToday.mockReturnValueOnce(true);

    const result = await loadProfileAsync();
    expect(result).toMatchSnapshot();
  });

  test('should return null for profile streak if it is not defined', async () => {
    getMultiAsyncData.mockReturnValueOnce(profileNoStreakResponse);

    const result = await loadProfileAsync();
    expect(result).toMatchSnapshot();
  });

  describe('error', () => {
    test('should throw an error', async () => {
      try {
        getMultiAsyncData.mockReturnValueOnce({ error: 'error' });
        await loadProfileAsync();
      } catch (err) {
        expect(err.name).toEqual('Error');
      }
    });
  });
});
