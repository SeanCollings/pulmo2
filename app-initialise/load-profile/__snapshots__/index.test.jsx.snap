// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadProfileAsync - unit test should not set streak to 0 if last activity date is same day 1`] = `
Object {
  "selected_level": 2,
  "streak": Object {
    "current": 2,
    "lastActivity": "2020-07-26T15:15:43.411Z",
  },
}
`;

exports[`loadProfileAsync - unit test should return null for profile streak if it is not defined 1`] = `
Object {
  "selected_level": 2,
}
`;
